File: src/favoringIsolatedMutability/java/coordtyped/UseAccountService.java
/***
 * Excerpted from "Programming Concurrency on the JVM",
 * published by The Pragmatic Bookshelf.
 * Copyrights apply to this code. It may not be used to create training material, 
 * courses, books, articles, and the like. Contact us if you are in doubt.
 * We make no guarantees that this code is fit for any purpose. 
 * Visit http://www.pragmaticprogrammer.com/titles/vspcon for more book information.
***/
package com.agiledeveloper.pcj;

import akka.actor.TypedActor;
import akka.actor.ActorRegistry;
import akka.actor.Actors;

public class UseAccountService {
  
  public static void main(final String[] args) 
    throws InterruptedException {

    final Account account1 = 
      TypedActor.newInstance(Account.class, AccountImpl.class);
    final Account account2 = 
      TypedActor.newInstance(Account.class, AccountImpl.class);
    final AccountService accountService = 
      TypedActor.newInstance(AccountService.class, AccountServiceImpl.class);

    account1.deposit(1000);
    account2.deposit(1000);

    System.out.println("Account1 balance is " + account1.getBalance());
    System.out.println("Account2 balance is " + account2.getBalance());

    System.out.println("Let's transfer $20... should succeed");

    accountService.transfer(account1, account2, 20);

    Thread.sleep(1000);

    System.out.println("Account1 balance is " + account1.getBalance());
    System.out.println("Account2 balance is " + account2.getBalance());

    System.out.println("Let's transfer $2000... should not succeed");
    accountService.transfer(account1, account2, 2000);

    Thread.sleep(6000);

    System.out.println("Account1 balance is " + account1.getBalance());
    System.out.println("Account2 balance is " + account2.getBalance());
    
    Actors.registry().shutdownAll();
  }
}

File: bin/concurrency-code/favoringIsolatedMutability/scala/coordtyped/UseAccountService.scala
package com.agiledeveloper.pcj

import akka.actor.TypedActor
import akka.actor.ActorRegistry
import akka.actor.Actors

//START:CODE
object UseAccountService {
  
  def main(args : Array[String]) = {
    val account1 = 
      TypedActor.newInstance(classOf[Account], classOf[AccountImpl])
    val account2 = 
      TypedActor.newInstance(classOf[Account], classOf[AccountImpl])
    val accountService = 
      TypedActor.newInstance(
        classOf[AccountService], classOf[AccountServiceImpl])

    account1.deposit(1000)
    account2.deposit(1000)
    
    println("Account1 balance is " + account1.getBalance())
    println("Account2 balance is " + account2.getBalance())


    println("Let's transfer $20... should succeed")

    accountService.transfer(account1, account2, 20)

    Thread.sleep(1000)

    println("Account1 balance is " + account1.getBalance())
    println("Account2 balance is " + account2.getBalance())

    println("Let's transfer $2000... should not succeed")
    accountService.transfer(account1, account2, 2000)

    Thread.sleep(6000)

    println("Account1 balance is " + account1.getBalance())
    println("Account2 balance is " + account2.getBalance())
    
    Actors.registry.shutdownAll
  }
}
//END:CODE
